### PR TITLE
MRP documents refactor

### PR DESCRIPTION
**Pull request by Dotni**
Description of the issue/feature this PR addresses:

In https://github.com/odoo/odoo/pull/132739, the introduction of product documents generalized
the previous document approach used in MRP.

This commit removes the existing mrp_document model and merge it's data and behaviour with the
product document one.

In addition, a field attached_to_mrp added on the product document  will allow the user to
access documents in spefic places, like in a production or a bill of materials.
This commit also adds those documents to the respective chatters of the productions and bill
of materials.

task-id: 3649991
closes: https://github.com/odoo/odoo/pull/152286
enterprise: https://github.com/odoo/enterprise/pull/55634
upgrade: https://github.com/odoo/upgrade/pull/5622

Current behavior before PR:

Desired behavior after PR is merged:

---
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

**Review by Whenrow — CHANGES_REQUESTED** [review]

File: addons/mrp/security/ir.model.access.csv [hunk 1]
```diff
@@ -46,8 +46,8 @@ access_product_tag_mrp_manager,product.tag.mrp.manager,product.model_product_tag
 access_resource_calendar_manufacturinguser,resource.calendar manufacturing.user,resource.model_resource_calendar,mrp.group_mrp_user,1,0,0,0
 access_mrp_unbuild,mrp.unbuild,model_mrp_unbuild,group_mrp_user,1,1,1,1
 access_mrp_unbuild_manager,mrp.unbuild manager,model_mrp_unbuild,group_mrp_manager,1,1,1,1
-access_mrp_document_mrp_manager,mrp.document group_user,model_mrp_document,group_mrp_manager,1,1,1,1
-access_mrp_document_mrp_user,mrp.document group_user,model_mrp_document,group_mrp_user,1,1,1,1
+access_mrp_document_mrp_manager,product.document group_user,model_product_document,group_mrp_manager,1,1,1,1
```
Comment (Whenrow) [inline addons/mrp/security/ir.model.access.csv:49]: I think you need to adapt this in your migration script as well

File: addons/mrp/models/mrp_bom.py [hunk 2]
```diff
@@ -476,6 +476,36 @@ def _update_order_line_info(self, product_id, quantity, child_field=False, **kwa
...
+            if is_byproduct:
+                product_domain = OR([product_subdomain, [
+                    '|',
+                    '&', ('res_model', '=', 'product.product'), ('res_id', '=', bom.byproduct_ids.product_id.id),
+                    '&', ('res_model', '=', 'product.template'), ('res_id', '=', bom.byproduct_ids.product_id.product_tmpl_id.id)]])
+            else:
+                product_domain = product_subdomain
+            prod_final_domain = AND([bom_domain, product_domain])
+            final_domain = OR([final_domain, prod_final_domain]) if final_domain else prod_final_domain
+
```
Comment (Whenrow) [inline addons/mrp/models/mrp_bom.py:504]: what about document on components (`move_raw_ids.product_id`) ?

File: addons/mrp/models/product_document.py [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+    attached_on_mrp = fields.Selection(
+        selection=[
+            ('hidden', "Hidden"),
+            ('bom', "Bill of Materials")
+        ],
+        required=True,
+        string="MRP : Visible at",
+        help="Leave empty if document only accessible on product form.\n"
+            "Select Bill of Materials to visualise this document as a product attachment when this product is in a bill of material.\n"
+            "Select Manufacturing Order and the document will be visible in the chatter of the production order.",
```
[the remark this answers is not in the document]
Comment (Dotni) [reply addons/mrp/models/product_document.py:22]: This was removed from the spec 

> - Remove Manufacturing Order visbility option for now

File: addons/mrp/models/mrp_bom.py (hunk 2 above)
Comment (Dotni) [reply addons/mrp/models/mrp_bom.py:504]: The ``_get_extra_attachments`` method is only used to display documents of the main template/product of the bom in the chatter, so I think this is right

**Comment by Whenrow** [comment]
robodoo r+

File: addons/mrp/models/product_document.py [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+        return "bom" if self.env.context.get('attached_on_bom') else "hidden"
+
+    attached_on_mrp = fields.Selection(
+        selection=[
+            ('hidden', "Hidden"),
+            ('bom', "Bill of Materials")
+        ],
+        required=True,
+        string="MRP : Visible at",
+        help="Leave empty if document only accessible on product form.\n"
```
Comment (KangOl) [inline addons/mrp/models/product_document.py:20]: Can't leave it empty, as it is required!

File: addons/mrp/models/product_document.py [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+from odoo import fields, models
+
+
+class ProductDocument(models.Model):
+    _inherit = 'product.document'
+
+    def _default_attached_on_mrp(self):
+        return "bom" if self.env.context.get('attached_on_bom') else "hidden"
+
+    attached_on_mrp = fields.Selection(
```
Comment (KangOl) [inline addons/mrp/models/product_document.py:13]: I don't understand why you don't reuse `attached_on`? A single document can be on a BoM *and* on a quotation?

Comment (Dotni) [reply addons/mrp/models/product_document.py:13]: The document indeed could be "linked" to multiple objects.

**Comment by Whenrow** [comment]
robodoo r+

**Pull request closed without merging**